src/generators/notebookgenerator/itemrenderer.tsx

```diff
@@ -13,8 +15,161 @@ import { sanitizerOptions } from '../shared';
 
 import * as React from 'react';
 
+/**
+ * Returns the header level for a provided cell.
+ *
+ * ## Notes
+ *
+ * -   If a cell does not contain a header, the function returns the sentinel value `-1`.
+ *
+ * -   Cell header examples:
+ *
+ *     -   Markdown header:
+ *
+ *         ```
+ *         # Foo
+ *         ```
+ *
+ *     -   Markdown header:
+ *
+ *         ```
+ *         Foo
+ *         ===
+ *         ```
+ *
+ *         ```
+ *         Foo
+ *         ---
+ *         ```
+ *
+ *     -   HTML heading:
+ *
+ *         ```
+ *         <h3>Foo</h3>
+ *         ```
+ *
+ * @private
+ * @param cell - notebook cell
+ * @returns header level
+ */
+function headerLevel(cell: Cell): number {
+  if (cell.constructor.name !== 'MarkdownCell') {
+    return -1;
+  }
+  const lines = cell.model.value.text.split('\n');
```

Comment:
This stuff is fiddly, and there are a lot of edge cases. I think we should refactor to use the same logic that is used to generate cell headings, or otherwise reuse the `INotebookHeading` for a cell.

Reply:
I followed the same general [logic](https://github.com/jupyterlab/jupyterlab-toc/blob/2a3291ae31ce6f63b470c6865e43f39e23e12b61/src/generators/notebookgenerator/index.ts#L411) as used to generate Markdown headings, so simply followed precedent.